pkg/gadgets/network-graph/gadget.go

```diff
@@ -0,0 +1,306 @@
+	t.enricher, err = NewEnricher(withKubernetes, trace.Spec.Node)
+	if err != nil {
+		trace.Status.OperationError = fmt.Sprintf("Failed to start network-graph enricher: %s", err)
+		return
+	}
+	t.tracer, err = nettracer.NewTracer(&nettracer.Config{})
+	if err != nil {
+		trace.Status.OperationError = fmt.Sprintf("Failed to start network-graph tracer: %s", err)
+		return
+	}
+
+	genKey := func(container *pb.ContainerDefinition) string {
+		if container.Netns == t.netnsHost {
+			return "host"
+		}
+		return container.Namespace + "/" + container.Podname
+	}
+
+	attachContainerFunc := func(container *pb.ContainerDefinition) error {
+		key := genKey(container)
+
+		err := t.tracer.Attach(key, container.Pid)
+		if err != nil {
+			t.publishMessage(trace, eventtypes.ERR, key, fmt.Sprintf("failed to attach tracer: %s", err))
+			return err
+		}
+		t.publishMessage(trace, eventtypes.DEBUG, key, "tracer attached")
+		return nil
+	}
+
+	detachContainerFunc := func(container *pb.ContainerDefinition) {
+		key := genKey(container)
+
+		err := t.tracer.Detach(key)
+		if err != nil {
+			t.publishMessage(trace, eventtypes.ERR, key, fmt.Sprintf("failed to detach tracer: %s", err))
+			return
+		}
+		t.publishMessage(trace, eventtypes.DEBUG, key, "tracer detached")
+	}
+
+	containerEventCallback := func(event pubsub.PubSubEvent) {
+		switch event.Type {
+		case pubsub.EventTypeAddContainer:
+			attachContainerFunc(&event.Container)
+		case pubsub.EventTypeRemoveContainer:
+			detachContainerFunc(&event.Container)
+		}
+	}
+
+	t.pubSubKey = pubSubKey(fmt.Sprintf("gadget/network-graph/%s/%s", trace.ObjectMeta.Namespace, trace.ObjectMeta.Name))
+	existingContainers := t.resolver.Subscribe(
+		t.pubSubKey,
+		*gadgets.ContainerSelectorFromContainerFilter(trace.Spec.Filter),
+		containerEventCallback,
+	)
+
+	for _, c := range existingContainers {
+		err := attachContainerFunc(c)
+		if err != nil {
+			log.Warnf("Warning: couldn't attach BPF program: %s", err)
+			break
+		}
+	}
+	t.started = true
+
+	trace.Status.State = "Started"
+
+	t.done = make(chan bool)
+	t.wg.Add(1)
+	go t.run(trace)
+}
+
+func (t *Trace) run(trace *gadgetv1alpha1.Trace) {
+	defer t.wg.Done()
+
+	traceBeforePatch := trace.DeepCopy()
+	ticker := time.NewTicker(time.Second)
+
+	events := []types.Event{}
+	for {
+		select {
+		case <-t.done:
+			ticker.Stop()
+			return
+		case <-ticker.C:
+			if t.tracer == nil {
+				// This should not happen with t.wg
+				log.Errorf("tracer is nil at tick")
+				t.publishMessage(trace, eventtypes.ERR, "host", "tracer is nil at tick")
+				return
+			}
+			newEdges, err := t.tracer.Pop()
+			if err != nil {
+				log.Errorf("failed to read BPF map: %s", err)
+				t.publishMessage(trace, eventtypes.ERR, "host", fmt.Sprintf("failed to read BPF map: %s", err))
+				return
+			}
+			newEvents := t.enricher.Enrich(newEdges)
+
+			events = append(events, newEvents...)
+			events = types.Unique(events)
+			trace.Status.Output = types.EventsString(events)
```

Comment:
I think it is strange that a Stream gadget also generates a sort of summary of the events on the `Status`. It is like mixing two output modes, don't you think? IIUC, in the `Status.Output` we are generating the network-graph with no repeated entries while we are showing all the network activity in the stream, right?

If my understanding is correct, I think we have two kinds of gadgets here, the stream one looks like a sort of tcpdump while the status shows the final network-graph. Would it make sense to separate them into two different gadgets? Otherwise, what about adding a parameter to tell the gadget when to generate a normal stream of events and when to generate just a summary but everything using the Stream output? Something like the `--timeout` of the top gadgets. WDYT?

I think keeping this mixed behaviour makes it difficult to write the documentation like I mentioned in previous comments and also could confuse users.

cc @mauriciovasquezbernal

Reply:
Good point. @alban do you have any use case in mind where storing this information in the CR is useful?

Reply:
For now, `kubectl-gadget` only uses the stream so I suggest to delete the code that updates the `Status.Output`.

The `Status.Output` field was useful for the GlobalTrace prototype where the network policies are generated in the Kubernetes cluster. But the GlobalTrace prototype is not merged, so it's fine to delete that code for now; we can always add it again if/when needed.

Reply:
It is just because the events are annotated with the local pod and remote pod/svc/endpoint so that it's possible to generate a graph showing the communication between pods.

I'm open for alternative names: naming is difficult :)